fdbserver/DataDistribution.actor.cpp

```diff
@@ -1335,6 +1343,140 @@ ACTOR Future<Void> bulkLoadingCore(Reference<DataDistributor> self, Future<Void>
+	return Void();
+}
+
+ACTOR Future<bool> scheduleBulkDumpTasks(Reference<DataDistributor> self) {
+	state Database cx = self->txnProcessor->context();
+
+	state Key beginKey = normalKeys.begin;
+	state Key endKey = normalKeys.end;
+	state KeyRange rangeToRead;
+
+	state int bulkDumpResultIndex = 0;
+	state BulkDumpState bulkDumpState;
+	state KeyRange bulkDumpRange;
+	state RangeResult bulkDumpResult;
+
+	state int rangeLocationIndex = 0;
+	state std::vector<IDDTxnProcessor::DDRangeLocations> rangeLocations;
+	state KeyRange taskRange;
+	state bool allComplete = true;
+
+	while (beginKey < endKey) {
+		state Transaction tr(cx);
+		try {
+			rangeToRead = Standalone(KeyRangeRef(beginKey, endKey));
+			bulkDumpResult.clear();
+			wait(store(
+			    bulkDumpResult,
+			    krmGetRanges(&tr, bulkDumpPrefix, rangeToRead, SERVER_KNOBS->DD_BULKDUMP_TASK_METADATA_READ_SIZE)));
+			bulkDumpResultIndex = 0;
+			for (; bulkDumpResultIndex < bulkDumpResult.size() - 1; bulkDumpResultIndex++) {
+				if (bulkDumpResult[bulkDumpResultIndex].value.empty()) {
+					continue;
+				}
+				bulkDumpRange = Standalone(
+				    KeyRangeRef(bulkDumpResult[bulkDumpResultIndex].key, bulkDumpResult[bulkDumpResultIndex + 1].key));
+				bulkDumpState = decodeBulkDumpState(bulkDumpResult[bulkDumpResultIndex].value);
+				if (bulkDumpState.getPhase() == BulkDumpPhase::Complete) {
+					continue;
+				}
+				ASSERT_WE_THINK(bulkDumpState.getPhase() == BulkDumpPhase::Submitted);
+				// Partition the job in the unit of shard
+				allComplete = false;
+				wait(store(rangeLocations, self->txnProcessor->getSourceServerInterfacesForRange(bulkDumpRange)));
+				rangeLocationIndex = 0;
+				for (; rangeLocationIndex < rangeLocations.size(); ++rangeLocationIndex) {
+					// Spawn task per shard
+					taskRange = rangeLocations[rangeLocationIndex].range;
+					ASSERT(!taskRange.empty());
+					if (!self->ongoingBulkDumpActors.liveActorAt(taskRange.begin)) {
+						// Limit parallelism
+						loop {
+							if (self->bulkDumpParallelismLimitor.tryIncrementTaskCounter()) {
+								break;
+							}
+							wait(self->bulkDumpParallelismLimitor.waitUntilCounterChanged());
+						}
+						// In case no ongoing task on the same range
+						SSBulkDumpTask task = getSSBulkDumpTask(rangeLocations[rangeLocationIndex].servers,
+						                                        bulkDumpState.spawn(taskRange));
+						// Issue task
+						self->ongoingBulkDumpActors.insert(
+						    taskRange,
+						    doBulkDumpTask(self, task.targetServer, task.bulkDumpState, task.checksumServers));
+					}
+					beginKey = rangeLocations.back().range.end;
+				}
+			}
+			beginKey = bulkDumpResult.back().key;
+			wait(delay(0.1));
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled) {
+				throw e;
+			}
+			wait(tr.onError(e));
+		}
+	}
+	return allComplete;
+}
+
+ACTOR Future<Void> bulkDumpTaskScheduler(Reference<DataDistributor> self) {
```

Comment:
Whats this doing?

Reply:
It monitors whether the all tasks completed by the `scheduleBulkDumpTasks`